feat(parser): parse regular expression with regex parser

Boshen opened the pull request:
Many false positives and incorrect errors. @leaysgur Enjoy :grin:

Run `just conformance` to update the snapshot.

Boshen commented on tasks/coverage/src/suite.rs:
```diff
@@ -329,6 +329,9 @@ pub trait Case: Sized + Sync + Send + UnwindSafe {
                 GraphicalReportHandler::new().with_theme(GraphicalTheme::unicode_nocolor());
             let mut output = String::new();
             for error in errors {
+                if error.message.starts_with("Invalid regular expression") {
```
Remove this if you wanna see all the errors.

leaysgur commented:
@Boshen Thanks for the setup! And I have a 1 question.

How should I know errors to fix? Check each snapshot like this...?

![image](https://github.com/user-attachments/assets/662acf19-f521-486c-ac8c-f35548be0b8f)

Boshen commented:
It's not producing the correct error for

```
Expect Syntax Error: tasks/coverage/test262/test/language/literals/regexp/named-groups/invalid-non-id-continue-groupspecifier-4-u.js
Expect Syntax Error: tasks/coverage/test262/test/language/literals/regexp/named-groups/invalid-non-id-continue-groupspecifier-4.js
Expect to Parse: tasks/coverage/test262/test/annexB/built-ins/RegExp/incomplete_hex_unicode_escape.js
```

e.g. https://github.com/tc39/test262/blob/main/test/language/literals/regexp/named-groups/invalid-non-id-continue-groupspecifier-4-u.js

have the comment 

```

/*---
description: GroupSpecifier must be identifier-like.
esid: prod-GroupSpecifier
negative:
  phase: parse
  type: SyntaxError
features: [regexp-named-groups]
---*/

$DONOTEVALUATE();

/(?<a\>.)/u;
```

---

Everything else are false positives 😅 they should not error.

leaysgur commented:
Thank you! I will try to fix them reported in `parser_test262.snap`.

- [ ] Expect Syntax Error: tasks/coverage/test262/test/language/literals/regexp/named-groups/invalid-non-id-continue-groupspecifier-4-u.js
- [ ] Expect Syntax Error: tasks/coverage/test262/test/language/literals/regexp/named-groups/invalid-non-id-continue-groupspecifier-4.js
- [ ] Expect to Parse: tasks/coverage/test262/test/annexB/language/literals/regexp/class-escape.js
- [ ] Expect to Parse: tasks/coverage/test262/test/built-ins/String/prototype/split/separator-regexp.js
- [x] Expect to Parse: tasks/coverage/test262/test/built-ins/RegExp/regexp-modifiers/**/*
  - ⚠️ Stage 3 Modifiers are not supported yet

`rg 'Expect (to Parse|Syntax Error)' parser_*.snap | rg regexp` showed nothing for other snapshots.

- parser_typescript
- parser_babel
- parser_misc

leaysgur commented:
@Boshen Is this what you are looking for...?!

![image](https://github.com/user-attachments/assets/b23fb425-b599-4499-b241-c3fd9280ee53)

Boshen commented:
yes yes yes!

> @Boshen Is this what you are looking for...?!
> 
> ![image](https://private-user-images.githubusercontent.com/6259812/359701634-b23fb425-b599-4499-b241-c3fd9280ee53.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyMjcxOTUsIm5iZiI6MTcyNDIyNjg5NSwicGF0aCI6Ii82MjU5ODEyLzM1OTcwMTYzNC1iMjNmYjQyNS1iNTk5LTQ0OTktYjI0MS1jM2ZkOTI4MGVlNTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjFUMDc1NDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjE3MGNlMjkxNmU0MGRiMmYyYmI1NmIxN2Q4NDNmYjY3ZjE5ZTBlYjZmNWE5ZDY5ZTMyZmZhZTdmMzM0YjBhNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.06veID_kvUKy6vJNt5ga4d_mXVOr9dXj0i-KT1afF7k)

The next batch seems to the ones in `tasks/coverage/parser_typescript.snap`

e.g. 
```
Expect Syntax Error: tasks/coverage/typescript/tests/cases/conformance/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInRegularExpressions08.ts
```

Boshen commented:
I ran this in our E2E test suite and got the following errors.

```
Error
"node_modules/.pnpm/ajv-formats@3.0.1_ajv@8.17.1/node_modules/ajv-formats/dist/formats.js"
  x Invalid regular expression: Too many capturing groups

Error
"node_modules/.pnpm/ajv-formats@3.0.1_ajv@8.17.1/node_modules/ajv-formats/src/formats.ts"
  x Invalid regular expression: Too many capturing groups

Error
"node_modules/.pnpm/ajv-formats@3.0.1_ajv@8.17.1/node_modules/ajv-formats/src/formats.js"
  x Invalid regular expression: Too many capturing groups

Error
"node_modules/.pnpm/@nuxt+devtools@1.3.9_rollup@4.21.0_supports-color@9.4.0_vite@5.4.1_@types+node@22.4.1_less@4._t2hmrj4los77odlqbefbkljiye/node_modules/@nuxt/devtools/dist/client/_nuxt/JOMs3kJS.js"
  x Invalid regular expression: Too many capturing groups

Error
"node_modules/.pnpm/ajv-formats@2.1.1_ajv@8.17.1/node_modules/ajv-formats/dist/formats.js"
  x Invalid regular expression: Too many capturing groups

Error
"node_modules/.pnpm/ajv-formats@2.1.1_ajv@8.17.1/node_modules/ajv-formats/src/formats.ts"
  x Invalid regular expression: Too many capturing groups

Error
"node_modules/.pnpm/ajv-formats@2.1.1_ajv@8.17.1/node_modules/ajv-formats/src/formats.js"
  x Invalid regular expression: Too many capturing groups
```

It's missing spans.

Boshen commented:
> These all seem to be errors due to the use of the u flag in target: es5.

I think we do not have such a mechanism, what should we do?

> Here too, in all cases, the RegExp are used but errors are not related to RegExp pattern itself.

Let's skip these.

Boshen commented:
> I will add Spans for Too many capturing groups tomorrow~!

They seem to be false positives too 😅

Boshen commented:
> I will add Spans for Too many capturing groups tomorrow~!

They seem to be false positives too 😅

leaysgur commented:
@Boshen 

- Ignore TS tests using `target:es5`: https://github.com/oxc-project/oxc/pull/4998/commits/67ddf091ab2aa83b6711294fdc3a256f1cd057a1
  - `tasks/coverage/typescript/tests/cases/conformance/parser/ecmascript5/RegularExpressions/` are left as is
  - These were bugs of oxc_parser itself from the start 👀 
- Fix false positive `Too many capturing groups`: https://github.com/oxc-project/oxc/pull/4998/commits/195ff4d31f256be6ca5b763b78f9328ebf0dfc0b
  - Just a mistake, `2 ^ 32` should be `2.pow(32)` in Rust...
- Add `Span`s for all error reports:  https://github.com/oxc-project/oxc/pull/4998/commits/353d4d6256acf02d1bf5ee2ac294d9e2f2952ba2

Boshen commented:
Nice work! 💯 

The last favor I wanna ask is whether you'd like to review all the code once more and see if there are any refactoring opportunities or performance improves that can be found.

I always do another round of reviews myself when I finish such a big project, because I often forget what I wrote in the first place 😅

Boshen commented:
My next task would be adding the regex to the AST and have it accessible in the linter.

Boshen committed:
feat(parser): parse regular expression with regex parser (#4998)

Many false positives and incorrect errors. @leaysgur Enjoy :grin:

Run `just conformance` to update the snapshot.

leaysgur commented:
✌🏻 

> I always do another round of reviews myself when I finish such a big project, because I often forget what I wrote in the first place 😅

Coincidence!
I’ve been working on something similar, and I already have some notes for that. 😎 

However, there are a few things I need to confirm with you, so I’ll create another issue later.
